### DISCUSSION TITLE
Getting Crash in failed assertion `mZLightNear >= mNear' for point light

### DISCUSSION THREAD

**Discussion by akshay-varshney** [google/filament#6102] [Q&A]
I am trying to add point light. This is the code to add point light:

`filament::LightManager::Builder(filament::LightManager::Type::POINT)
                    .color(filament::Color::toLinear<filament::ACCURATE>(filament::sRGBColor(0.98f, 0.92f, 0.89f)))
                    .intensity(200000.f, filament::LightManager::EFFICIENCY_LED)
                    .position({ lightDirection0.x()*1000, lightDirection0.y()*1000, lightDirection0.z()*1000 })
                    .falloff(1200)
                    .castShadows(true)
                    .build(*mEngine, mSunLight0);`

for setting this is the code:
` 
 auto& lcm = mEngine->getLightManager();
    lcm.setPosition(lcm.getInstance(mSunLight), {lightDirection0.x()*1000, lightDirection0.y()*1000, lightDirection0.z()*1000});
    lcm.setIntensity(lcm.getInstance(mSunLight0), lightIntensity0);
    lcm.setFalloff(lcm.getInstance(mSunLight0), 1200);

    filament::LinearColor lightColor0 = {0.98f, 0.92f, 0.89f};
    
    lcm.setColor(lcm.getInstance(mSunLight0), lightColor0);`

I am getting crash, Can someone please help what are we doing wrong?

**Comment by pixelflinger** [comment]
The default `mZLightNear ` is set at 5m. I'm guessing your camera's near plane is larger than that. 
You simply need to adjust the `mZLightNear ` using `View::setDynamicLightingOptions()`.

**Comment by akshay-varshney** [comment]
@pixelflinger : I tried with setDynamicLightingOptions and set a very far manual value like this `theView->setDynamicLightingOptions(5000, 15000);`  but still it is crashing. Can you please help what values we need to set ? Also does we are using both cameras orthographic and perspective. 

This is crashing for both spot light and point light. Any help on this one?

**Reply by pixelflinger** [reply]
What near and far plane are you using for your camera?

**Reply by akshay-varshney** [reply]
@pixelflinger 
this is how we are calculating
  ```
  mNear = std::max(0.1, std::abs(near));
    mFar = std::abs(far) ;
```

We have set camera properties in our code according to the sun light now we are using same for spot light.

do we need to make any changes in camera properties to make it work for spot light?

**Reply by pixelflinger** [reply]
What are the actual values for your camera near and far?

I don't understand what it means to set the camera properties according to the sun light.
The camera near/far should not depend on your light. It should only depend on the scene.

**Reply by akshay-varshney** [reply]
@pixelflinger This is how we are calculating near and far 
```
 near = std::numeric_limits<double>::infinity();
  far = -std::numeric_limits<double>::infinity();
    auto fun = [&](Eigen::Vector3d b, const Eigen::Matrix3d& rotation)
    {
        if (rotation != Eigen::Matrix3d::Identity())
            b = rotation * b;
        near = std::min(near, mCameraDistance - b[2]);
        far = std::max(far, mCameraDistance - b[2]);
        b[2] -= mCameraDistance;
    };
```

The camera near/far should not depend on your light. It should only depend on the scene. -- Got it

**Reply by pixelflinger** [reply]
I thought I was asking something simple. What are the NUMERICAL VALUES of near/far you are passing to the Camera object. Just print it and tell me what it is.

This should be set only once and for all when you initialize your scene.

In your code here, I don't know what b, rotation and mCameraDistance is.

**Comment by akshay-varshney** [comment]
@pixelflinger Sorry, please find the values of 
```
near =  21838.901563293908
far = 22061.703831798735
 mNear= 21837.901563293908
mZLightFar = 100
mZLightNear	= 5
```

**Reply by pixelflinger** [reply]
Well as you can see, the zLightNear is smaller than your near, you must adjust it to be always larger.


That being said, I don't know what your application is, but these near/far values seem very very unusual.

Typically values for near/far are 0.1/100 if your units are meters (10cm to 100m).
There is usually no reason for the near plane to be so far from the camera origin. These values only depend on what's in your scene and are usually set once and never changed.

**Reply by bejado** [reply]
@pixelflinger Maybe we should clamp `zLightNear` to always be >= the camera near to avoid this assertion firing.

**Reply by pixelflinger** [reply]
Yes, I need to make sure there aren't other implications.

**Comment by akshay-varshney** [comment]
@pixelflinger : Somehow we figured out the reason for crash and it is no longer crashing but cannot see point light on 3D object.
This is the builder code which we have written for point light.

```
 filament::LightManager::Builder(filament::LightManager::Type::POINT)
                       .color(filament::Color::toLinear<filament::ACCURATE>(filament::sRGBColor(0.0f, 0.0f, 1.0f)))
                       .intensity(2000.0f, filament::LightManager::EFFICIENCY_LED)
                       .position({20.0f, 20.0f, 20.0f })
                       .falloff(50.0f)
                       .build(*mEngine, mSunLight0);
```

Can you please suggest what might be going wrong here? Any insights?

**Reply by romainguy** [reply]
It depends on how far the light is from your object, what your camera exposure is, etc.
